### PR TITLE
Issue 122 new relational

### PR DESCRIPTION
**Pull request by n1v0lg**
This adds functionality for input/ output for collections (lists and matrices), conditional selections and swapping, oblivious permutations, shuffling, and aggregation. 

Addresses #122

### CODE REVIEW

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/AdvancedNumeric.java [hunk 1]
```diff
@@ -127,6 +129,20 @@
...
+   *
+   * @param condition the Computation holding the condition on which to select.
+   * @param left the Computation holding the left argument.
+   * @param right the Computation holding the right argument.
+   * @return an computation holding the output of the appended protocol.
+   */
+  DRes<SInt> condSelect(DRes<SInt> condition, DRes<SInt> left,
+      DRes<SInt> right);
+  
+  DRes<Pair<DRes<SInt>, DRes<SInt>>> condSwap(DRes<SInt> condition, DRes<SInt> left,
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/AdvancedNumeric.java]: It would be nice to have javadoc here

Comment (pffrandsen) [reply core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/AdvancedNumeric.java]: An idea: Maybe it makes more sense to coin these two method:
- selectIf
- swapIf

?

File: core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+  private final Matrix<BigInteger> openMatrix;
+  private final int inputParty;
+  private final int h;
+  private final int w;
+  private final boolean isInputProvider;
+
+  /**
+   * Makes a new CloseMatrix.
+   *
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java]: Try to run the checkstyle on your classes:
- mvn checkstyle:checkstyle

I doubt it will be graceful on this javadoc.

File: core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        // get neighbor element of output element
+        OElIdx = neighElIdxOf(OElIdx);
+
+        /*
+         * note that we are guaranteed to terminate since we are picking the next element to visit
+         * based on a permutation so the only element we are ever going to revisit is the one we
+         * started with at which point we stop
+         */
+        if (OElIdx == firstOElIdx) {
+          break;
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java]: Please, could we drop the while true and remove this goto call?

File: core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        }
+      }
+    }
+
+    /*
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java]: There is a lot of comments in this method, which is hard to read - could we split it up in sub methods (named as stated in the comment) and get easier to read code?

File: core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java [hunk 5]
```diff
@@ -165,23 +163,22 @@ private boolean checkDimensions(LPTableau tableau, Matrix<DRes<SInt>> updateMatr
           return seq.par((par) -> {
             ArrayList<DRes<SInt>> nextBasis = new ArrayList<>(noConstraints);
             for (int i = 0; i < noConstraints; i++) {
-              nextBasis.add(par.seq(
-                  new ConditionalSelect(exitingIndex.get(i), ent, state.basis.get(i))));
+              nextBasis.add(
+                  par.seq(new ConditionalSelect(exitingIndex.get(i), ent, state.basis.get(i))));
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java]: Strange realignment - why did you move this?

File: core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java [hunk 6]
```diff
@@ -154,9 +153,8 @@ private boolean checkDimensions(LPTableau tableau, Matrix<DRes<SInt>> updateMatr
    */
   private DRes<LPState> phaseTwoProtocol(ProtocolBuilderNumeric builder, LPState state) {
     return builder.seq((seq) -> seq.seq(
-        new ExitingVariable(state.tableau, state.updateMatrix, state.enteringIndex, state.basis))
-    ).pairInPar(
-        (seq, exitingVariable) -> {
+        new ExitingVariable(state.tableau, state.updateMatrix, state.enteringIndex, state.basis)))
+        .pairInPar((seq, exitingVariable) -> {
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java]: Strange realignment - why did you move this?

File: core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java [hunk 7]
```diff
@@ -165,23 +163,22 @@ private boolean checkDimensions(LPTableau tableau, Matrix<DRes<SInt>> updateMatr
...
-                exitingVariable.updateColumn, state.pivot, state.prevPivot))
-    ).seq((seq, pair) -> {
-      List<DRes<SInt>> basis = pair.getFirst();
-      state.updateMatrix = pair.getSecond();
-      state.basis = basis;
-      // // Copy the resulting new update matrix to overwrite the current
-      state.prevPivot = state.pivot;
-      return () -> state;
-    });
+        }, (seq, exitingVariable) -> seq
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java]: Strange realignment - why did you move this?

File: core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java [hunk 8]
```diff
@@ -196,19 +193,20 @@ private boolean checkDimensions(LPTableau tableau, Matrix<DRes<SInt>> updateMatr
...
-      state.terminationOut = seq.numeric().open(positive);
-      state.enteringIndex = entering;
-      return () -> state;
-    });
+    return builder
+        .seq(
+            // Compute potential entering variable index and corresponding value of
+            // entry in F
+            new EnteringVariable(state.tableau, state.updateMatrix))
+        .seq((seq, enteringAndMinimum) -> {
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java]: Strange realignment - why did you move this?

File: core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java [hunk 9]
```diff
@@ -196,19 +193,20 @@ private boolean checkDimensions(LPTableau tableau, Matrix<DRes<SInt>> updateMatr
...
-      List<DRes<SInt>> entering = enteringAndMinimum.getFirst();
-      SInt minimum = enteringAndMinimum.getSecond();
-      // Check if the entry in F is non-negative
-      DRes<SInt> positive = seq.comparison().compareLEQLong(zero, () -> minimum);
-      state.terminationOut = seq.numeric().open(positive);
-      state.enteringIndex = entering;
-      return () -> state;
-    });
+    return builder
+        .seq(
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java]: Strange realignment - why did you move this?

File: core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java [hunk 10]
```diff
@@ -221,19 +219,19 @@ private boolean checkDimensions(LPTableau tableau, Matrix<DRes<SInt>> updateMatr
...
-        new BlandEnteringVariable(state.tableau, state.updateMatrix)
-    ).seq((seq, enteringAndMinimum) -> {
-      List<DRes<SInt>> entering = enteringAndMinimum.getFirst();
-      SInt termination = enteringAndMinimum.getSecond();
-      state.terminationOut = seq.numeric().open(() -> termination);
-      state.enteringIndex = entering;
-      return () -> state;
-    });
+  private DRes<LPState> blandPhaseOneProtocol(ProtocolBuilderNumeric builder, LPState state) {
+    return builder
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java]: Strange realignment - why did you move this?

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/Collections.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+  public DRes<Matrix<DRes<SInt>>> closeMatrix(Matrix<BigInteger> openMatrix, int inputParty);
+
+  public DRes<Matrix<DRes<SInt>>> closeMatrix(int h, int w, int inputParty);
+
+  public DRes<Matrix<DRes<BigInteger>>> openMatrix(DRes<Matrix<DRes<SInt>>> closedMatrix);
+
+  // Conditional
+
+  public DRes<List<DRes<SInt>>> condSelect(DRes<SInt> condition, DRes<List<DRes<SInt>>> left,
```
Comment (Rhaea) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/Collections.java]: Javadoc missing on all these wonderful functionalities. Would love to know exactly what they did.

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/Collections.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+  public DRes<Matrix<DRes<SInt>>> permute(DRes<Matrix<DRes<SInt>>> values, int permProviderPid);
+
+  public DRes<Matrix<DRes<SInt>>> shuffle(DRes<Matrix<DRes<SInt>>> values, Random rand);
+
+  public DRes<Matrix<DRes<SInt>>> shuffle(DRes<Matrix<DRes<SInt>>> values);
+
+  // Relational (SQL-like) operators
+
+  public DRes<Matrix<DRes<SInt>>> leakyAggregateSum(DRes<Matrix<DRes<SInt>>> values, int groupColIdx,
```
Comment (Rhaea) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/Collections.java]: Especially javadoc on this one. Leaky does not sound good, so I bet most ppl would love to know exactly what they leak.

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/AdvancedNumeric.java (hunk 1 above)
Comment (n1v0lg) [reply core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/AdvancedNumeric.java]: Agree with swapIf, don't like selectIf. To me selectIf implies that something will be selected if the condition is true and nothing else will happen otherwise.

File: core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java (hunk 2 above)
Comment (n1v0lg) [reply core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java]: You are right it was not. The entire code base has over 4000 formatting warnings. Perhaps we can have a big reformatting pass soon?

File: core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java (hunk 6 above)
Comment (n1v0lg) [reply core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java]: All the realignments because I ran auto-format on those files.

File: core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java (hunk 3 above)
Comment (n1v0lg) [reply core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java]: Done.

File: core/src/main/java/dk/alexandra/fresco/lib/collections/LinearLookUp.java [hunk 13]
```diff
@@ -26,13 +26,14 @@
  *******************************************************************************/
 package dk.alexandra.fresco.lib.collections;
 
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/LinearLookUp.java]: Organize Imports

File: core/src/main/java/dk/alexandra/fresco/lib/collections/MatrixUtils.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
+package dk.alexandra.fresco.lib.collections;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/MatrixUtils.java]: Organize Imports

File: core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseList.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.stream.Collectors;
+
+public class CloseList implements ComputationParallel<List<DRes<SInt>>, ProtocolBuilderNumeric> {
+
+  private final List<BigInteger> openInputs;
+  private final int numberOfInputs;
+  private final int inputParty;
+  private final boolean isInputProvider;
+
+  public CloseList(List<BigInteger> openInputs, int inputParty) {
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseList.java]: JavaDoc

File: core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseList.java [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    super();
+    this.openInputs = openInputs;
+    this.numberOfInputs = openInputs.size();
+    this.inputParty = inputParty;
+    this.isInputProvider = true;
+  }
+
+  public CloseList(int numberOfInputs, int inputParty) {
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseList.java]: JavaDoc

File: core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+public class CloseMatrix
+    implements ComputationParallel<Matrix<DRes<SInt>>, ProtocolBuilderNumeric> {
+
+  private final Matrix<BigInteger> openMatrix;
+  private final int inputParty;
+  private final int height;
+  private final int width;
+  private final boolean isInputProvider;
+
+  public CloseMatrix(Matrix<BigInteger> openMatrix, int inputParty) {
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java]: JavaDoc

File: core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    super();
+    this.openMatrix = openMatrix;
+    this.height = openMatrix.getHeight();
+    this.width = openMatrix.getWidth();
+    this.inputParty = inputParty;
+    this.isInputProvider = true;
+  }
+
+  public CloseMatrix(int h, int w, int inputParty) {
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/io/CloseMatrix.java]: JavaDoc

File: core/src/main/java/dk/alexandra/fresco/lib/collections/permute/PermuteRows.java [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+
+  private final DRes<Matrix<DRes<SInt>>> values;
+  private final int[] idxPerm;
+  private final int permProviderPid;
+  private final boolean isPermProvider;
+  private final WaksmanUtils wutils;
+  // not final as this will be set during protocol execution
+  private Matrix<DRes<SInt>> cbits;
+
+  public PermuteRows(DRes<Matrix<DRes<SInt>>> values, int[] idxPerm, int permProviderPid,
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/permute/PermuteRows.java]: This really needs some JavaDoc. What does those arguements do? :)

File: core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+
+  public int[] invert(int[] p) {
+    int[] inv = new int[p.length];
+    for (int i = 0; i < p.length; i++) {
+      inv[p[i]] = i;
+    }
+    return inv;
+  }
+
+  public Matrix<BigInteger> setControlBits(int[] perm) throws UnsupportedOperationException {
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java]: JavaDoc

File: core/src/main/java/dk/alexandra/fresco/lib/compare/gt/LessThanOrEquals.java [hunk 21]
```diff
@@ -33,7 +33,8 @@
 import dk.alexandra.fresco.framework.builder.numeric.ProtocolBuilderNumeric;
 import dk.alexandra.fresco.framework.util.Pair;
 import dk.alexandra.fresco.framework.value.SInt;
-import dk.alexandra.fresco.lib.compare.ConditionalSelect;
+import dk.alexandra.fresco.lib.conditional.ConditionalSelect;
+
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/compare/gt/LessThanOrEquals.java]: No space

File: core/src/main/java/dk/alexandra/fresco/lib/lp/BlandEnteringVariable.java [hunk 22]
```diff
@@ -23,6 +23,10 @@
  */
 package dk.alexandra.fresco.lib.lp;
 
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/BlandEnteringVariable.java]: Organize imports

File: core/src/main/java/dk/alexandra/fresco/lib/lp/OptimalValue.java [hunk 23]
```diff
@@ -31,6 +31,8 @@
 import dk.alexandra.fresco.framework.builder.numeric.AdvancedNumeric;
 import dk.alexandra.fresco.framework.builder.numeric.ProtocolBuilderNumeric;
 import dk.alexandra.fresco.framework.value.SInt;
+import dk.alexandra.fresco.lib.collections.Matrix;
+
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/lib/lp/OptimalValue.java]: no space
